### PR TITLE
Use subjects to bind textfields

### PR DESCRIPTION
**Pull request by Odrakir**
## What's this?
Binds textfields to subjects instead of using RxAction

### Relationships
Doesn't close any issue.

## Why do we want this?
This started as a way to stop using `RxAction.Inputs` in the VCs, but in the end @khoi found out that using RxActions with textfields could be problematic. So instead I'm repurposing this PR to change RxAction into Subjects for textfields.

## How is it done?
It uses `BehaviorSubject` to get the textfield values. In the case of `TokenResetViewModel` there are some more changes in other for it to work.

## :squid: Permissions
Go ahead.

### CODE REVIEW

**Comment by Odrakir** [comment]
Fair enough. Let’s do that. Although... `BindAction` does work flawlessly when the action itself is an Observable (as opposed to a `async Task`)

File: Toggl.Giskard/Activities/SelectClientActivity.cs [hunk 1]
```diff
@@ -39,8 +39,8 @@ protected override void OnCreate(Bundle bundle)
                 .BindAction(ViewModel.Close)
                 .DisposedBy(DisposeBag);
 
-            filterEditText.Rx().Text()
-                .Subscribe(ViewModel.SetFilterText.Inputs)
+            filterEditText.Rx()
```
Comment (kornelijepetak) [inline Toggl.Giskard/Activities/SelectClientActivity.cs]: I see changes to Giskard's bindings and no extension method on Giskard.

**Comment by Odrakir** [comment]
I agree with @khoi , sort of...

I'm just refactoring the call in places where we are already using an action so we don't have to access the `Inputs` property. The code was already using RxAction. I'm not trying to refactor the VM itself, which we probably should. 

I see two cases here:

- Binding an `UITextField` to some property in the VM, in which case we are better of using a `BehaviorRelay` or some other kind of `Subject`
- Binding an `UITextField` so it executes some operation on the VM (like searching for something while typing), in which case it makes more sense to use a `RxAction`

But now, after @khoi mentioned this, I took a look at these specific changes and all the actions are doing is setting values in subjects, so yes. This shouldn't be actions at all. Which doesn't mean a UITextField's text shouldn't ever be bound to an action.

**Comment by khoi** [comment]
I agree that the behavior of the app doesn't change because of this PR since we're already binding the text to an action's input directly.

Refactor is due for binding those events to a Subject, or a Relay. Merging this PR would standardize these type of bindings.

> Binding an UITextField so it executes some operation on the VM (like searching for something while typing), in which case it makes more sense to use a RxAction

Binding TextField's text changed directly to a RxAction without somehow involving `Enabled` will cause 💥 sooner or later. 

In your example, I think it'd safer to do this

```
class VM {
   var TextRelay: BehaviorRelay<string>  
   var Output = TextRelay.Debounce(0.3).SelectMany(...etc)
}
class View {
   TextField.Bind(ViewModel.TextRelay).Disposeby(bag)
   ViewModel.Output.BindToSomeView().Disposeby(bag)
}
```

**Comment by khoi** [comment]
> I don't know why you say that. The code you posted can be done with just one action property.

pls collaborate, cause I think it would crash if you use one Action property

**Comment by Odrakir** [comment]
@khoi Is right, so I repurposed this PR to use Subjects for textfields instead of RxActions. I've updated the description.

File: Toggl.Daneel/Extensions/Reactive/UITextFieldExtensions.cs [hunk 2]
```diff
@@ -3,6 +3,7 @@
 using System.Reactive.Disposables;
 using System.Reactive.Linq;
 using Toggl.Foundation.MvvmCross.Reactive;
+using Toggl.Multivac.Extensions;
```
Comment (simonrozsival) [inline Toggl.Daneel/Extensions/Reactive/UITextFieldExtensions.cs]: 👀

File: Toggl.Foundation.MvvmCross/ViewModels/SelectClientViewModel.cs [hunk 3]
```diff
@@ -27,16 +27,12 @@ public sealed class SelectClientViewModel : MvxViewModel<SelectClientParameters,
         private long workspaceId;
         private long selectedClientId;
         private SelectableClientViewModel noClient;
-        private ISubject<string> clientFilterText = new BehaviorSubject<string>(string.Empty);
 
         public IObservable<IEnumerable<SelectableClientBaseViewModel>> Clients { get; private set; }
-
+        public readonly ISubject<string> FilterText = new BehaviorSubject<string>(string.Empty);
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectClientViewModel.cs]: Don't use public fields, instead use a property
```suggestion
        public ISubject<string> FilterText { get; } = new BehaviorSubject<string>(string.Empty);
```

File: Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs [hunk 4]
```diff
@@ -20,15 +20,11 @@ public sealed class SelectCountryViewModel : MvxViewModel<long?, long?>
...
-        private readonly ISubject<string> filterText = new BehaviorSubject<string>(string.Empty);
-
         private readonly IMvxNavigationService navigationService;
 
-        public InputAction<SelectableCountryViewModel> SelectCountry { get; }
-
         public IObservable<IEnumerable<SelectableCountryViewModel>> Countries { get; private set; }
-
-        public InputAction<string> SetFilterText { get; }
+        public readonly ISubject<string> FilterText = new BehaviorSubject<string>(string.Empty);
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs]:
```suggestion
        public ISubject<string> FilterText { get; } = new BehaviorSubject<string>(string.Empty);
```

File: Toggl.Foundation.MvvmCross/ViewModels/TokenResetViewModel.cs [hunk 5]
```diff
@@ -30,28 +29,23 @@ public sealed class TokenResetViewModel : MvxViewModel
...
         public IObservable<bool> IsPasswordMasked { get; }
-
         public IObservable<bool> HasError { get; }
         public IObservable<string> Error { get; }
-
-        public IObservable<bool> IsLoading { get; }
-
         public IObservable<bool> NextIsEnabled { get; }
 
+        public readonly BehaviorSubject<string> Password = new BehaviorSubject<string>(string.Empty);
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/ViewModels/TokenResetViewModel.cs]:
```suggestion
        public BehaviorSubject<string> Password { get; } = new BehaviorSubject<string>(string.Empty);
```

Comment (simonrozsival) [reply Toggl.Foundation.MvvmCross/ViewModels/TokenResetViewModel.cs]: Also why is it explicitly a `BehaviorSubject` and not just `ISubject`? Do you use some behavior-subject specific properties somewhere? I guess in this case it's OK and it communicates that the last value is returned on subscription. I guess the `Take(1)` is the reason.

File: Toggl.Foundation.MvvmCross/ViewModels/TokenResetViewModel.cs [hunk 6]
```diff
@@ -161,24 +136,21 @@ private async Task signout()
         }
 
         private IObservable<Unit> done() =>
-            Observable.Create<Unit>(observer =>
-            {
-                if (!passwordSubject.Value.IsValid)
+            Password
+                .Take(1)
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/ViewModels/TokenResetViewModel.cs]: I would like `FirstAsync` better, but they do the same I guess.

File: Toggl.Foundation.MvvmCross/ViewModels/TokenResetViewModel.cs [hunk 7]
```diff
@@ -161,24 +136,21 @@ private async Task signout()
...
+                .Select(Multivac.Password.From)
+                .Select(password =>
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/ViewModels/TokenResetViewModel.cs]: We have a `ThrowIf(password => !password.IsValid, new InvalidOperationException())` extension method which would be perfect for this piece of code.

File: Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs [hunk 8]
```diff
@@ -57,7 +52,8 @@ public override async Task Initialize()
                 allCountries.Insert(0, selectedElement);
             }
 
-            Countries = filterText
+            Countries = FilterText
+                .Select(text => text ?? string.Empty)
                 .Select(text => text.Trim())
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs]: You can save one line of code:
```suggestion
                 .Select(text => text?.Trim() ?? string.Empty)
```

File: Toggl.Foundation.MvvmCross/ViewModels/TokenResetViewModel.cs (hunk 6 above)
Comment (kornelijepetak) [reply Toggl.Foundation.MvvmCross/ViewModels/TokenResetViewModel.cs]: They are not the same.
If sequence emits no elements, `Take(1)` will continue, `FirstAsync` will throw.

Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/ViewModels/TokenResetViewModel.cs]: In this case `FirstAsync` makes more sense. Thanks.

**Pull request merged**